src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+
+    public EnsemblBiomartRetriever(String retrieverName)
+    {
+        super(retrieverName);
+    }
+
+    private static final int UNIPROT_QUERY_TOTAL = 2;
+    private static int successCount;
+    private static final String  microarrayTypesBaseQuery = "type=listAttributes&mart=ENSEMBL_MART_ENSEMBL&virtualSchema=default&dataset=BIOMART_SPECIES_NAME_gene_ensembl&interface=default&attributePage=feature_page&attributeGroup=external&attributeCollection=microarray";
+
+    /**
+     * Downloads Ensembl-Microarray and Ensembl-Uniprot identifier mapping files for all species, if they exist.
+     * @throws Exception
+     */
+    protected void downloadData() throws Exception {
+
+        // TODO: Add unit tests
+        // Get names of all organisms we add links and/or microarray data for.
+        // Species names are in biomart format (eg: hsapiens).
+        for (String biomartSpeciesName : getBiomartSpeciesNames()) {
+            logger.info("Retrieving Biomart files for " + biomartSpeciesName);
+            // Create directory where Biomart files will be stored.
+            new File(this.destination).mkdirs();
+
+            logger.info("Retrieving microarray data");
+            // Query Biomart for existing microarray 'types' (not ids) that exist for this species.
+            Set<String> microarrayTypes = queryBiomart(getMicroarrayTypesQuery(biomartSpeciesName), 0);
+            // Iterate through each microarray type and retrieve Ensembl-Microarray identifier mappings.
+            // All mappings are stored in a single file, (eg: hsapiens_microarray);
+            retrieveAndStoreMicroarrayData(biomartSpeciesName, getBiomartXMLPath(), microarrayTypes);
+
+            // Query Ensembl-Uniprot (swissprot and trembl) identifier mapping data from Biomart
+            // and write it to a file (eg: hsapiens_uniprot).
+            logger.info("Retrieving UniProt data");
+            retrieveAndStoreUniprotData(biomartSpeciesName, getBiomartXMLPath());
+
+            logger.info("Completed Biomart data retrieval for " + biomartSpeciesName);
+        }
+    }
+
+    /**
+     * Query Biomart for both Uniprot mapping types (curated -- SwissProt, generated -- TrEMBL).
+     * @param biomartSpeciesName
+     * @throws Exception
+     */
+    private void retrieveAndStoreUniprotData(String biomartSpeciesName, String biomartXMLPath) throws IOException {
+        Set<String> uniprotIdLines = new HashSet<>();
+        try {
+            for (String uniprotQueryId : Arrays.asList("uniprotswissprot", "uniprotsptrembl")) {
+                // If proper response, write data to file.
+                Set<String> uniprotResponseLines = queryBiomart(getBiomartIdentifierQuery(biomartXMLPath, biomartSpeciesName, uniprotQueryId), 0);
+                for (String uniprotResponseLine : uniprotResponseLines) {
+                    List<String> tabSplit = Arrays.asList(uniprotResponseLine.split("\t"));
+                    if (tabSplit.size() == 4) {
```

Comment:
This condition is a bit cryptic.  It might be better to move this to a private helper method and call it something like this: `if (containsUniProtIdentifier(tabSplit))`.  You could then have the method body be:

```
private boolean containsUniProtIdentifier(List<String> tabSplit) {
    final int UNIPROT_COLUMN = 3;

    String uniProtIdentifier = tabSplit.get(UNIPROT_COLUMN);
    return uniProtIdentifier != null && !uniProtIdentifier.trim().isEmpty();
}
```

Reply:
It turns out you still get the `ArrayOutOfBoundsException` here, even if checking for null. Based on what I've been able to find, it seems like checking the array size is the way to go. I'll give the necessary integer a proper name though to denote it. 

https://examples.javacodegeeks.com/java-basics/exceptions/java-lang-arrayindexoutofboundsexception-how-to-handle-array-index-out-of-bounds-exception/
https://www.journaldev.com/18092/java-lang-arrayindexoutofboundsexception

Reply:
Okay, sounds good.  I think I'm still in favour of having a named method to explain what is being tested, though.

Maybe the method body could be:

```
private boolean containsUniProtIdentifier(List<String> tabSplit) {
    final int UNIPROT_COLUMN = 3;

    return tabSplit.size() > UNIPROT_COLUMN &&  
           tabSplit.get(UNIPROT_COLUMN) != null &&
           !tabSplit.get(UNIPROT_COLUMN).trim().isEmpty();
}
```

Reply:
Yep I'd added them to their own named method, though your way only uses one variable so I will use that